### ISSUE TITLE
Making UDFs work for other Office Application is painful

### ISSUE THREAD

**Issue by IRBorisov** [xlwings/xlwings#1621]
#### OS (e.g. Windows 10 or macOS Sierra)
Windows 10

#### Versions of xlwings, Excel and Python (e.g. 0.11.8, Office 365, Python 3.7)
xlwings 0.23.3
Excel 2016 (64 bit)
Word 2016 (64 bit)
Python 3.9.5 (64 bit)

#### Describe your issue (incl. Traceback!)
I intend to use xlwings as middle layer connecting out VBA and Python tools for text manipulation. For that i needed to run Python scripts from Word VBA.
I followed docs on how to run UDF from other (than Excel) office applications ([relevant docs page](https://docs.xlwings.org/en/stable/other_office_apps.html))

**I had to jump through some hoops to make it work and I hope this description will help improving xlwings to make this procedure easier in future releases.**
Further description highlights changes i had to implement **not documented** in xlwings docs.
1. After I imported xlwings and xlwings_udfs into Word I tried compiling it and failed due to some Excel-specific code
1.1. In xlwings.bas I commented out functions: sql, has_dynamic_array
1.2. Replace everywhere "Application.Hwnd" -> "Application.ActiveWindow.Hwnd" (may not work for other apps, but Word and Excel are just fine with that change)
1.3. Replace "Range" -> "Excel.Range" (Word has its own Range object). I suggest qualifying all Office model objects with appropriate namespace.
1.4. In function XLPyCommand comment out block under 'If ThisWorkbook.name <> ThisWorkbook.name' and replace it with PYTHONPATH = ThisDocument.Path & ";" & GetConfig("PYTHONPATH") - this works only for Word and Visio. Need a better solution here.
2. For some reason command generated by XLPyCommand using global Python path was truncating path to python to the first space (i.e. 'C:\Program'). Quoting whole path didn't work out, but for some reason quoting only Program Files works (might be some cmd.exe quirk for my installation). In case you have similar issues, try adding this code to XLPyCommand
```
PYTHON_WIN = VBA.LCase(PYTHON_WIN)
PYTHON_WIN = VBA.Replace(PYTHON_WIN, "program files (x86)", """program files (x86)""")
PYTHON_WIN = VBA.Replace(PYTHON_WIN, "program files", """program files""")
```
3. After I successfully launched Python from VBA I got this trace of an error on loading my UDF module:
```python
xlwings server running, clsid={D62330A9-7971-42FA-B577-FB554D16B819}, event loop on 6216
pythoncom error: Python error invoking COM method.

Traceback (most recent call last):
  File "C:\Program Files\Python39\lib\site-packages\win32com\server\policy.py", line 278, in _Invoke_
    return self._invoke_(dispid, lcid, wFlags, args)
  File "C:\Program Files\Python39\lib\site-packages\win32com\server\policy.py", line 283, in _invoke_
    return S_OK, -1, self._invokeex_(dispid, lcid, wFlags, args, None, None)
  File "C:\Program Files\Python39\lib\site-packages\win32com\server\policy.py", line 586, in _invokeex_
    return func(*args)
  File "C:\Program Files\Python39\lib\site-packages\xlwings\server.py", line 198, in CallUDF
    res = call_udf(script, fname, args, this_workbook, FromVariant(caller))
  File "C:\Program Files\Python39\lib\site-packages\xlwings\udfs.py", line 423, in call_udf
    module = get_udf_module(module_name, this_workbook)
  File "C:\Program Files\Python39\lib\site-packages\xlwings\udfs.py", line 387, in get_udf_module
    wb = Book(impl=xlplatform.Book(Dispatch(xl_workbook)))
  File "C:\Program Files\Python39\lib\site-packages\win32com\client\__init__.py", line 96, in Dispatch
    return __WrapDispatch(dispatch, userName, resultCLSID, typeinfo, clsctx=clsctx)
  File "C:\Program Files\Python39\lib\site-packages\win32com\client\__init__.py", line 43, in __WrapDispatch
    return dynamic.Dispatch(dispatch, userName, WrapperClass, typeinfo, clsctx=clsctx)
  File "C:\Program Files\Python39\lib\site-packages\win32com\client\dynamic.py", line 122, in Dispatch
    typeinfo = IDispatch.GetTypeInfo()
AttributeError: 'NoneType' object has no attribute 'GetTypeInfo'
```
4. After some time watching through the 'xlwings\udfs.py' I found the reason. Apparently on Line 377 Python tries to find loaded module by name (in case it was already loaded), after that on line 389 it tries to look into workbook for embedded code module and crashes because it is called without a valid workbook caller. **I suggest implementing a check before trying to look for embedded code**
5. As a mitigation strategy I commented out the whole embedded code part (see attached udfs.py) and was able to run my UDF smoothly from VBA Word :-)

I feel like it should be way easier to use xlwings to query Python API from all Office Application and I hope my case description will help improving on the product.

Thanks to xlwings contributors and keep up the good work!

#### Include a minimal code sample to reproduce the issue (and attach a sample workbook if required!)
In attachments archive:
- udfs.py - modified xlwings distribution file, see lines 386-392
- xlwings.conf - config file
- standalone2.xlsm - excel standalone with working UDF (execute macro Module1.Test)
- standalone2.py - sample python code to run as UDF
- standalone2.docm - word standalone with working UDF if changes to udfs are applied (see traceback for error log) (execute macro Module1.Test). Also see changes in VBA code in module "xlwings"
[standalone2.zip](https://github.com/xlwings/xlwings/files/6669230/standalone2.zip)

**Comment by fzumstein** [comment]
agreed on it being painful - this was more of a POC that never got productionized as I didn't see too much response on it...

**Comment by IRBorisov** [comment]
Well it works just fine for me in the end.
I ditched all of the VBA API and most of the config settings except for PYTHONPATH and PYTHON_WIN.

After wrapping Python process into VBA Class and caching Py Object I was able to achieve 900 ms response time for simple function call vs 500 ms response from RPC to another Office Application through regular VBA model. Probably can be further optimized, but im happy with the result.

Now I'm researching the most effective way to update cached python object in VBA when python process dies before office process. Probably have to modify xlwings.dll code for that to add another Job to notify Caller process.
